wiki/Oppia-Android-Code-Coverage.md

```diff
@@ -0,0 +1,375 @@
+
+This section appears at the bottom of the report, as a drop-down. It includes a table listing the exemptions as,
+
+1. **File:** Displays the file name that is exempted (Clicking the drop-down reveals the file's path in the codebase).
+2. **Failure Reason:** Describes the specific reason for its exemption.
+
+With this report, you can review the coverage percentages for various files and identify which files pass or fail the coverage check.
+
+
+## 2. Local Command-Line Interface (CLI) Tools
+
+While the CI check provides an overview of coverage, you might want to visualize how tests cover specific files locally, including which lines are covered and which are not. For this, Oppia's local command-line coverage tooling is useful.
+
+Note: Follow these [Bazel setup instructions](https://github.com/oppia/oppia-android/wiki/Oppia-Bazel-Setup-Instructions) if Bazel isn't yet set up on your device:
+
+Oppia Android allows you to generate coverage reports in HTML format using the command:
+
+### Run Coverage
+
+```
+bazel run //scripts:run_coverage -- <path_to_root> <list_of_relative_path_to_files>
+```
+
+- <path_to_root>: Your root directory.
+- <list_of_relative_path_to_files>: Files you want to generate coverage reports for.
+
+For example, to analyze coverage for the file MathTokenizer.kt, use the relative path:
+
+```
+bazel run //scripts:run_coverage -- $(pwd) utility/src/main/java/org/oppia/android/util/math/MathTokenizer.kt
+```
+
+By default, this will generate an HTML report in the coverage_reports directory. For the given file, the report will be saved as **coverage_reports/utility/src/main/java/org/oppia/android/util/math/MathTokenizer/coverage.html**
+
+A list of files can be provided as an input to generate coverage reports for all the files.
+
+```
+bazel run //scripts:run_coverage -- $(pwd) 
+utility/src/main/java/org/oppia/android/util/math/MathTokenizer.kt 
+utility/src/main/java/org/oppia/android/util/parser/math/MathModel.kt
+```
+
+### Process Timeout
+
+The default wait time for a process to complete is 300 seconds (5 minutes). If the process does not finish within this period, you may encounter a "Process Did Not Finish Within Intended Time" error. To extend the wait time, you can modify the timeout setting by adding the flag --processTimeout=15 (The Time unit is minutes).
+
+```
+bazel run //scripts:run_coverage -- $(pwd)
+utility/src/main/java/org/oppia/android/util/parser/math/MathModel.kt --processTimeout=15
+```
+
+## 2.1 Understanding the CI Coverage Report
+
+Let's examine the coverage.html report located at coverage_reports/utility/src/main/java/org/oppia/android/util/math/MathTokenizer/coverage.html to understand its contents.
+
+**Header Section:** At the top of the report, you'll find information including:
+
+![image](https://github.com/user-attachments/assets/08c4f0a0-51b5-4d8e-a2f3-ff13d8b6cb2f)
+
+#
```

Comment:
Did you mean for these (here & below) to be headers? I think this section needs some formatting work since it looks a bit awkward right now in Markdown mode.

Reply:
By Header, I was trying to reference the top summary section. I think I cluttered the section.
But still not sure how to name the 2 portions, for now referenced them as:
- Coverage Overview
- Line-by-Line Coverage Breakdown

and tried reformatting it, does it look less cluttered or need further work?

Reply:
I think it looks good; I'll add additional suggestions for reorganizing if I think of any.